### DISCUSSION TITLE
A Problem in Ti parent grain reconstruction

### DISCUSSION THREAD

**Discussion by huseyinhzl** [ORTools4MTEX/ORTools#51] [Q&A] [answered]
Hi,

I am trying to reconstruct parent beta grains in Ti6Al4V alloy, yet I am having errors that I could not find a way out. I have tried voting and variant graph algorithms until now, but both were inconclusive with regard to grain reconstruction. The errors that I got are as follows:

1. If I calculate the child grains with 1.5*degree, I am getting strange lines after reconstructing the clean parent grains as shown below. I think this is emerging due to the small wild spikes in the reconstructed grains. 

![Lines in grains](https://github.com/ORTools4MTEX/ORTools/assets/120036622/ab57975b-0608-456a-afe1-e9cd3e6a43d9)
**Figure 1 - Parent beta grains with vertical lines**

2. When I clean the wild spikes less than 3 pixels and recalculate the grains,  I get a strange orientation plot, especially with the Burgers' relationship. The below image shows the disorientation graph that I got from my script. 

![Parent-Child Misorientation](https://github.com/ORTools4MTEX/ORTools/assets/120036622/242f86fa-d5d6-4e7a-9ae0-7e6250f0b27e)
**Figure 2 - Parent-Child grain Disorientation plot**

3. With no small grains, I did not see those strange vertical lines as in Figure 1, yet the corresponding parent grains seem not correct. Figure after job.calcParentFromVote command does not match with the clean reconstructed parent grains. I can also barely see grains from child maps. I am putting the figures after job.calcParentFromVote command and reconstructed parent grains with child map for comparison. Also, it seems there is some overfitting of parent grains in Figure 4.

![Voting](https://github.com/ORTools4MTEX/ORTools/assets/120036622/939b3e67-321b-4f6f-b39f-743a5b0db644)
**Figure 3 - Grain plot after job.calcParentFromVote command **

![ET100](https://github.com/ORTools4MTEX/ORTools/assets/120036622/38fae351-5d33-4596-ab38-542f8d18c8b7)
**Figure 4 - Reconstructed parent grains with child map**

My question is if my parent grain reconstruction is wrong. If so, how can I manage to fix it? I have several maps to be post-processed for parent grain reconstruction.

**Comment by AzdiarGazder** [comment]
Hi @huseyinhzl,

Not a worry - happy to help.

Since your figures are not labelled, it is difficult to point out exactly where your problems originate off-the-bat.
I am assuming the second figure is your original map of child grains.
If this is the case, the large white areas of zero solutions most probably are one of the main sources of error.

In any case, could you please upload the EBSD map data (as-mapped without cleaning) and the script(s)?
Also, could you please edit your query to:
- itemise your issues in the order of their appearance, and 
- label your figures.

The above would really help in diagnosing exactly what is going on.

Warm regards,
Azdi

**Reply by huseyinhzl** [reply]
Hi @AzdiarGazder,
Thank you so much for your reply. I have updated the question and hope my question looks better now. I am also attaching the original ctf file and my script as a ZIP file here.

[Original Data.zip](https://github.com/ORTools4MTEX/ORTools/files/12590952/Original.Data.zip)

**Reply by AzdiarGazder** [reply]
Hi @huseyinhzl,

Thank you for itemising your issues in the order of their appearance, and for labeling your figures.
It makes things a lot clearer to a third person like me who is not involved with your work.

And you're very welcome to the functions in mtexTools and ORTools.
As acknowledgement, please cite the use of these function libraries in the references section of your publications.

Now, to answer your questions:

1. As I mentioned in my earlier reply to you, one of the main issues in your map is the large amount of white areas comprising zero solutions. To resolve this issue, I imported your *.ctf file into OI Aztec Crystal and filled-in areas with zero solution(s). 
 
2. Following this, I had a look at the number of pixels making up each grain (under "Grain Size" in OI Aztec Crystal). Unfortunately, almost 85-90% of your map comprises grains with less than 5 pixels each. This means your 3D printed microstructure of child hcp Ti-Alpha grains is quite fine but the step size you used during map acquisition is very coarse.

Taking points 1 and 2 together means you have poor map quality leading to poor statistics. This is why you are facing issues during reconstruction in MTEX/ORTools. In an ideal world, I would advise you to re-do your maps using a finer step size but I understand that this may not be possible due to economic and/or beam time limitations.

So to help resolve your issues with the map data you have, here is the workflow:

- Import your *.ctf file into OI Aztec Crystal and fill-in areas with zero solution(s). Save and use this *.ctf file for parent grain reconstruction.
- Copy your filled in *.ctf file to C:\Users\YOUR_NAME\Documents\MATLAB\GitHub\ORTools\data\input\ebsd
- Use the following script in ORTools:
```matlab

% *********************************************************************
%                        ORTools - Ti6Al4V_ET100_C
% *********************************************************************
% Reconstruction of beta parent grains from
% alpha in titanium alloys.
% The MTEX description for reconstruction is given here:
% https://mtex-toolbox.github.io/TiBetaReconstruction.html
% https://www.youtube.com/watch?v=e6R0dApUc8Q&t=921s
% *********************************************************************
% Dr. Azdiar Gazder, 2020, azdiaratuowdotedudotau
% Dr. Frank Niessen, 2020, contactatfniessendotcom
% (Remove "dot" and "at" to make this email address valid)
% *********************************************************************
home; close all; clear variables;
currentFolder;
set(0,'DefaultFigureWindowStyle','normal');
screenPrint('StartUp','ORTools - Ti6Al4V_ET100_C');
%% Initialize MTEX
% Startup and set some settings
startup_mtex;
setMTEXpref('xAxisDirection','east');
setMTEXpref('zAxisDirection','outOfPlane');
setMTEXpref('FontSize',14);
setInterp2Tex;
% Default directories - Do not modify
Ini.dataPath = [strrep(pwd,'\','/'),'/data/'];
Ini.cifPath = [Ini.dataPath,'input/cif/'];
Ini.ebsdPath = [Ini.dataPath,'input/ebsd/'];
Ini.texturePath = [Ini.dataPath,'output/texture/'];
Ini.imagePath = [Ini.dataPath,'output/images/'];

%% Load data
% Load an MTEX dataset into 'ebsd'
% create an EBSD variable containing the data
ebsd = EBSD.load([Ini.ebsdPath,'Ti6Al4V_ET100_C.ctf'],'interface','ctf','convertSpatial2EulerReferenceFrame');

%% Compute, filter and smooth grains
screenPrint('SegmentStart','Computing, filtering and smoothing grains');
% Grains are calculated with a 3° threshold
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree,'removeQuadruplePoints');
% EBSD data in small grains are removed
ebsd(grains(grains.grainSize < 3)) = [];
% Recalculate the grains from the remaining data ...
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'angle',3*degree);
% ... and smooth the grain boundaries
grains = smooth(grains,5);

%% Rename and recolor phases
screenPrint('SegmentStart','Renaming and recoloring phases');
phaseNames = {'Alpha','Beta'};
% Rename "Ti (BETA) to "Beta"and "Ti (alpha)" to "Alpha"
ebsd = renamePhases(ebsd,phaseNames);
% Choose your favourite colors
ebsd = recolorPhases(ebsd);
%% Finding the orientation relationship
screenPrint('SegmentStart','Finding the orientation relationship(s)');
% Choose Beta as a parent and Alpha as a child phase
job = parentGrainReconstructor(ebsd,grains,Ini.cifPath);
job.p2c = orientation.Burgers(job.csParent, job.csChild);
% Check the disorientation
plotHist_OR_misfit(job,orientation.Burgers(job.csParent,job.csChild),...
                   'legend',{'Burgers OR'});
xlim([0,10]);
%     - The misfit with Burgers OR and the experimental OR is almost
%     identical
%% Plotting (with ORTools functions)
screenPrint('SegmentStart','Plotting some ORTools maps');
% Use some of the ORTools functions to visualize the determined OR
% and its relation to the microstructure

% Phase map
plotMap_phases(job,'linewidth',1);
%       - The microstructure consists of 99.75 % alpha

% Parent and child IPF maps
plotMap_IPF_p2c(job,vector3d.Z,'linewidth',1,'child');
%       - We can visually recognize the prior beta grains

% Child-child grain boundary misorientation map
plotMap_gB_c2c(job,'linewidth',1);
%       - Alpha of same prior beta grain seem to have ~58° misorientation

% Plot a map of the OR boundary disorientation, or misfit
plotMap_gB_misfit(job,'linewidth',1.5, 'maxColor', 10);
%       - A threshold of 10° shows where the prior beta boundaries are

% Plot parent-child and child-child OR boundary probability map
plotMap_gB_prob(job,'linewidth',1.5);
%       - This is also manifested in the OR probability map

% We can plot the OR and boundary misorientation axes and color the
% experimental points according to their disorientation angle from the OR
% We plot the
%       - parent-child misorientation axes in the parent basis
%       - parent-child misorientation axes in the child basis
%       - and the child-child misorientation axes in the child basis
plotIPDF_gB_misfit(job);

% And we can do the same for the OR probability
plotIPDF_gB_prob(job);
%% Compute parent orientations from triple junctions
% We can use a voting algorithm which votes for a parent orientation at
% triple points of child grains
job.calcTPVotes('minFit',2.5*degree,'maxFit',5*degree);
% Check the votes for all grains
figure
plot(job.grains,job.votes.prob(:,1));
mtexColorbar
% .... and calculate parent orientations for all grains with a > 70%
% probability
job.calcParentFromVote('minProb',0.7);
%Plot the data
figure;
plot(job.parentGrains,job.parentGrains.meanOrientation,'linewidth',1.5);
%% Grow parent grains at grain boundaries by voting algorithm
% We can then let the parent grains grow into the child grains by a voting
% algorithm
for k = 1:3
    job.calcGBVotes('p2c','threshold',k*2.5*degree);
    job.calcParentFromVote
end

% This is the resulting reconstructed parent microstructure
figure;
plot(job.parentGrains,job.parentGrains.meanOrientation,'linewidth',1.5);
%% Clean reconstructed grains
% We now merge grains with similar orientation
job.mergeSimilar('threshold',5*degree);
% and merge small inclusions into larger grains
job.mergeInclusions('maxSize',5);
% and then plot the cleaned reconstructed parent microstructure
figure;
plot(job.parentGrains,job.parentGrains.meanOrientation,'linewidth',1.5)
%% Variant analysis
% Plot the variant pole figure
figure;
plotPDF_variants(job);
% We can calculate variants and packets
job.calcVariants;
% and plot the variant map
plotMap_variants(job,'linewidth',3);
%plotMap_variants(job,'grains','linewidth',3); %Plot grain data instead
%% Plot reconstructed parent EBSD orientations and the IPF key
parentIPFkey = plotMap_IPF_p2c(job,vector3d.Z,'linewidth',3,'parent');
figure; plot(parentIPFkey);

return
%% Save images
saveImage(Ini.imagePath);

%% Check the beta grains interactively by clicking on them
grainClick(job,'noScalebar','noFrame');
%grainClick(job,'grains','noScalebar','noFrame'); %Plot grain data instead

```
- The result you get in plot number 17 is as follows:

![Ti6Al4V_ET100_C](https://github.com/ORTools4MTEX/ORTools/assets/28549013/72a7a694-685a-4aa1-a7c3-6331b0147ae7)

Hope this helps.
**If the above explanations have resolved your issues, please mark this message as your answer.**

Warm regards,
Azdi

**Reply by huseyinhzl** [reply]
Hi @AzdiarGazder ,

Thank you for your effort and help in my problem. I am aware that the step size that I used is not the best option for my alloy, yet due to the beam time and number of user doing EBSD I cannot use fine step sizes like 0.25um or less. I have still questions that is not clear in my mind. 
1. Why does reconstructed parent grain boundaries not follow the boundaries that we can see after voting algorithm (Figure 3 in my question)? Is it because of merging of the similar grains or due to the poor EBSD map? Also, AztecCrystal can do the parent grain analysis now, which is not perfect, but I am getting a map as below. When I do the parent grain calculation to compare it with MTEX version, I saw it was different as can be seen in the below. It might be misleading , I do not know.

![Ti6Al4V ET100-ParentGrains-IPF + GB](https://github.com/ORTools4MTEX/ORTools/assets/120036622/6fdfe81d-4197-4fa6-86be-c7bcee073ef6)
**Figure 1 - Parent grains calculated by AztecCrystal**

2. Though we are merging and cleaning the data, why do we still get some small grains within big prior beta grains? Are they the grain boundaries belonging to the residual child alpha phase?

Thank you again and I am sorry if I am asking too much

**Reply by AzdiarGazder** [reply, marked as answer]
Hi @huseyinhzl 

1. I advise you to please read up on the different reconstruction algorithms used in MTEX and OI Aztec Crystal - as the theory and application of each of are different.  Consequently, each of them should and will return different answers in terms of details. So, once you've done your due diligence finding out which algorithm is most applicable to your case, you should use it judiciously; ensuring that you are able to justify the application of the value of all the various variables.

2. In keeping with point 1, this is my second suggestion. Rather than run the script I included in my previous message as a whole, please go through it in a line-by-line manner. Compare and contrast what is happening to your map data after each line is executed. There is a lot of processing undertaken on your child grains before reconstruction - maybe one of those steps could be improved on or removed? This is also a part of the due diligence that you need to undertake for your data.

3.  I cannot state with 100% certainty  that you have small grains within big prior beta grains. They could very well be there on account of your 3D printing technique - but the poor quality of child grain data, in terms of both step size and indexing rate, means that there will always be doubt. There is always going to be a little uncertainty whether you've cleaned too much (or too little) before reconstruction. If the child grains were manipulated too much before reconstruction, then the parent grains are less representative of the as-mapped child grains. On the other hand, if If the child grains were manipulated too little before reconstruction, then the parent grains are a product of all the artifacts are already contained in the child map.

In summary, what I'm saying (in a very roundabout manner) is that unfortunately there is no magic bullet.
And especially so in cases when the map statistics are unreliable. 

Hope this helps.
**If any of above explanations have resolved your issues, please mark that message as your answer.**

Warm regards,
Azdi
